openstacksdk_plugin/resources/compute/server.py

```diff
@@ -577,11 +669,231 @@ def _update_server_config_from_relationships(server_config):
+
+    for net_id in networks:
+        if net_id not in attached_networks:
+            ctx.logger.info('Attaching network {0}...'.format(net_id))
+            interface = \
+                openstack_resource.create_server_interface({'net_id': net_id})
+            ctx.logger.info(
+                'Successfully attached network {0} to device (server) id {1}.'
+                .format(net_id, openstack_resource.resource_id))
+            added_interfaces.append(interface.id)
+        else:
+            ctx.logger.info(
+                'Skipping network {0} attachment, because it is already '
+                'attached to device (server) id {1}.'
+                .format(net_id, openstack_resource.resource_id))
+
+    # Check if there are interfaces added to the external server and add
+    # them as runtime properties
+    if added_interfaces:
+        ctx.instance.runtime_properties[SERVER_INTERFACE_IDS] =\
+            added_interfaces
+
+    # Set runtime properties for external server
+    server = openstack_resource.get()
+    _set_server_ips_runtime_properties(server)
+
+
+def _connect_resources_to_external_server(openstack_resource):
+    """
+    This method will try to connect resources to external server
+    :param openstack_resource: Instance Of OpenstackServer in order to
+    use it
+    """
+
+    # Try to connect networks to external server
+    _connect_networks_to_external_server(openstack_resource)
+
+    # Validate external key pair connected to the external server
+    _connect_keypair_to_external_server(openstack_resource)
+
+
+def _disconnect_resources_from_external_server(openstack_resource):
+    """
+    This method will disconnect networks from external server so that they
+    can be removed without any issue
+    :param openstack_resource: Instance Of OpenstackServer in order to
+    use it
+    """
+    # Delete all interfaces added to the external server
+    if SERVER_INTERFACE_IDS in ctx.instance.runtime_properties:
+        interfaces = ctx.instance.runtime_properties.get(
+            SERVER_INTERFACE_IDS, [])
+        for interface in interfaces:
+            openstack_resource.delete_server_interface(interface)
+            ctx.logger.info(
+                'Successfully detached network {0} to device (server) id {1}.'
+                .format(interface, openstack_resource.resource_id))
+
+
+def _assign_server_payload_as_runtime_properties(payload):
+    """
+    Store server configuration in the runtime
+    properties and cleans any potentially sensitive data.
+    :param payload: The payload.
+    """
+    if getattr(ctx, 'instance') and payload:
+        if SERVER_OPENSTACK_TYPE not in ctx.instance.runtime_properties.keys():
+            ctx.instance.runtime_properties[SERVER_OPENSTACK_TYPE] = {}
+        for key, value in payload.items():
+            if key not in ['user_data', 'adminPass']:
+                ctx.instance.runtime_properties[SERVER_OPENSTACK_TYPE][key]\
+                    = value
```

Comment:
Do we know that they always return JSON-serializable types?

Reply:
The good thing about openstacksdk it is returns all response if there is any as instance of dict which can be always retrieved easily and iterate over it